### PR TITLE
feat: two-mode incremental fetching (window + lookback) for Hive

### PR DESCRIPTION
**Pull request by matyas-jirat-keboola**
## What

Ports the CFTL-771 **two-mode incremental fetching** (default `watermark` + optional `incrementalFetchingLookback`, or absolute `window` via `incrementalFetchingStart`/`End`) to `ex-db-hive`. Already shipped in `ex-db-pgsql`; this is the hive port.

Hive is a low-code case — it uses the adapter's `DefaultQueryFactory` unchanged over ODBC, so the mode-aware `WHERE` comes for free from the released adapter. Hive only needs the opt-in hook + tests.

## Changes

- **`src/Extractor/Hive.php`** — add `getIncrementalFetchingColumnType()`, which opts Hive into the window/lookback feature by returning the resolver basetype of the incremental-fetching column. `validateIncrementalFetching()` is refactored to reuse a shared column-lookup helper; its raw-type allow-list check and its exact error message are **unchanged** (the existing unit test still asserts them).
  - **DATE columns:** `GenericStorage` maps a Hive `DATE` column to the `"DATE"` basetype, which the shared window/lookback resolver does not support (it handles `TIMESTAMP`/`INTEGER`/`NUMERIC`/`FLOAT`). Hive's comparison of a `DATE` column against the resolver's `Y-m-d H:i:s` literal is unverified, so the hook returns `null` for `DATE` — window/lookback then cleanly reports *"not supported by this extractor"* for a `DATE` column, rather than risk silently-wrong rows. Plain watermark on a `DATE` column is unchanged. `DATE` window/lookback can be enabled later once the comparison is verified against the live harness.
- **`composer.json` / `composer.lock`** — the two-mode feature lives in the shared libs, which **Phase 0 has not released yet** (no `1.17.0`/`17.3.0` tags). So this pins the three libs to the `cftl-771-incremental-fetching-window` branches via `repositories` VCS entries, exactly as `ex-db-pgsql` currently does. The adapter was the exact pin `1.15.2`. **These must be repinned to `config ^1.17`, `adapter ^1.17`, `common ^17.3` (dropping the VCS repos) once Phase 0 tags them — do not merge before then.**
- **Datadir tests** (all pass in CI against the live hive2 harness). Window mode:
  - `incremental-window-late-commit/` — `timestamp_col`, absolute start: a row below the stored watermark but inside the window still appears (watermark ignored).
  - `incremental-window-start-end/` — `timestamp_col`, absolute start **and** end (bounded range).
  - `incremental-window-relative/` — `timestamp_col`, **relative** start (`-100 years`): exercises the `strtotime` relative-bound path.
  - `incremental-window-int/` — **numeric** window on the integer `id` column (numeric `start`/`end`).
  - Lookback (watermark mode, no `incrementalFetchingMode`):
  - `incremental-lookback-timestamp/` — **timestamp** lookback (`26 hours`): exercises the resolver's timestamp date-arithmetic branch.
  - `incremental-lookback-late-commit/` — **FLOAT** numeric lookback.
  - `incremental-lookback-int/` — **INTEGER** numeric lookback (the common auto-increment case).
  - In each lookback/window case a row below the watermark but inside the bound re-appears, while a row beyond it stays excluded; all pre-existing `incremental-*` dirs still pass (backward compat).
- **`README.md`** — documents `incrementalFetchingMode` / `Lookback` / `Start` / `End` (+ examples); notes window/lookback is not supported for `DATE` columns.
- **`docker-compose.yml`** — *(separate `ci:` commit)* repin the LDAP test image `osixia/openldap:1.3.0-amd64` → `osixia/openldap:1.3.0`. Docker Hub removed the arch-specific tag (`manifest unknown`), which was breaking `docker compose build` for the whole repo's CI. Same OpenLDAP version, multi-arch tag, no config change.

## Verification

- CI green: `composer ci` = validate + phplint + **phpcs + phpstan (level max) + phpunit + datadir**, all against the live hive2 (LDAP) harness. The datadir suite includes the three new dirs **and** all pre-existing `incremental-*` dirs (backward compat).
- Backward compatible: default mode is `watermark`; the new schema nodes carry no default; existing configs are unchanged. The hook is only consulted when window/lookback is configured, so plain watermark (incl. on `DATE` columns) is untouched.

## Notes for reviewers

- **hive has no automated in-platform gate.** Unlike the other core DB extractors, hive's CI has no `tests-in-kbc` job and pushes a fixed `:test` image tag. The end-to-end functional proof here is the **datadir suite** (it runs window + lookback against a real Hive DB in the harness). Worth adding an automated in-platform gate for hive as a follow-up.
- Deps are intentionally on dev branches (see above) until the Phase 0 release; **do not merge** before the shared libs are released and the pins are updated.
- `DATE` window/lookback support is deliberately deferred (returns "not supported") until Hive's DATE literal comparison is verified — a good follow-up if there's demand.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/Extractor/Hive.php [hunk 1]
```diff
@@ -93,6 +87,38 @@ public function validateIncrementalFetching(ExportConfig $exportConfig): void
...
+     * only understands TIMESTAMP/INTEGER/NUMERIC/FLOAT, so DATE is normalised to TIMESTAMP here (the
+     * resolver emits "Y-m-d H:i:s", which Hive accepts when comparing a DATE column). Every other type
+     * in self::INCREMENTAL_TYPES already maps to a supported basetype.
+     */
+    public function getIncrementalFetchingColumnType(ExportConfig $exportConfig): ?string
+    {
+        $datatype = new GenericStorage($this->getIncrementalFetchingColumn($exportConfig)->getType());
+        $basetype = $datatype->getBasetype();
+
+        return $basetype === BaseType::DATE ? BaseType::TIMESTAMP : $basetype;
```
Comment (matyas-jirat-keboola) [inline src/Extractor/Hive.php]: **[#1 · Nice-to-have]** DATE columns are normalised to TIMESTAMP so the shared resolver emits a `Y-m-d H:i:s` literal, but Hive's comparison of a DATE column against that timestamp-formatted string is unverified and no datadir test covers a DATE column (this PR's tests use a timestamp and a numeric column, per the plan). Could we either add a DATE window/lookback datadir test to confirm real Hive behaviour, or return `null` for DATE (clean "not supported by this extractor" path) until it's verified? Plain watermark on a DATE column is unaffected either way.

**Review by matyas-jirat-keboola — COMMENTED** [review]
Automated component-review-pr pass (round 1, babysit loop). The change is backward-compatible and additive: new opt-in incremental-fetching params carry no schema default, and the getIncrementalFetchingColumnType() hook is only consulted when window/lookback is configured, so plain-watermark behaviour is unchanged. No secrets (test configs use %env(...)% placeholders). phpstan (level max), phpcs and the full datadir suite (incl. the two new dirs + all pre-existing incremental-* dirs) are green in CI. 3 nice-to-have findings inline. Blast radius is bounded regardless of config count because nothing existing changes shape.

File: src/Extractor/Hive.php [hunk 2]
```diff
@@ -93,6 +87,38 @@ public function validateIncrementalFetching(ExportConfig $exportConfig): void
...
+     * relative/absolute bounds). Reuses the same metadata lookup as validateIncrementalFetching().
+     *
+     * GenericStorage maps a Hive DATE column to the "DATE" basetype, but the window/lookback resolver
+     * only understands TIMESTAMP/INTEGER/NUMERIC/FLOAT, so DATE is normalised to TIMESTAMP here (the
+     * resolver emits "Y-m-d H:i:s", which Hive accepts when comparing a DATE column). Every other type
+     * in self::INCREMENTAL_TYPES already maps to a supported basetype.
+     */
+    public function getIncrementalFetchingColumnType(ExportConfig $exportConfig): ?string
+    {
+        $datatype = new GenericStorage($this->getIncrementalFetchingColumn($exportConfig)->getType());
```
Comment (matyas-jirat-keboola) [inline src/Extractor/Hive.php:106]: **[#2 · Nice-to-have]** For a window/lookback run the table metadata is fetched twice — `validateIncrementalFetching()` and `getIncrementalFetchingColumnType()` each call `getIncrementalFetchingColumn()->getTable()`, and neither caches. Consider memoising the resolved column/type. Low impact and identical to the released pgsql pattern, so only if it's cheap.

File: tests/functional/incremental-window-late-commit/source/data/config.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    "table": {
+      "tableName": "incremental",
+      "schema": "default"
+    },
+    "outputTable": "in.c-main.incremental-window-late-commit",
+    "primaryKey": ["id"],
+    "incremental": true,
+    "incrementalFetchingColumn": "timestamp_col",
+    "incrementalFetchingMode": "window",
+    "incrementalFetchingStart": "2020-01-05 05:00:00"
```
Comment (matyas-jirat-keboola) [inline tests/functional/incremental-window-late-commit/source/data/config.json:19]: **[#3 · Nice-to-have]** The window test exercises only `incrementalFetchingStart`; the `incrementalFetchingEnd` upper bound and the absolute-window-end warning guard aren't covered. Consider a case with both start and end asserting the bounded row set.

File: src/Extractor/Hive.php (hunk 2 above)
Comment (matyas-jirat-keboola) [reply src/Extractor/Hive.php:106]: Declining (verified independently): the metadata lookup runs twice per *run* (a startup cost, not per-row), and this non-memoized pattern is identical to the already-released reference driver. Memoizing here would diverge from it for negligible gain. Resolving.

File: composer.json [hunk 4]
```diff
@@ -11,9 +25,9 @@
         "jdorn/sql-formatter": "^1.2",
         "keboola/common-exceptions": "^1.0",
         "keboola/csv": "^3.3",
-        "keboola/db-extractor-adapter": "1.15.2",
-        "keboola/db-extractor-common": "^17.1",
-        "keboola/db-extractor-config": "^1.15",
+        "keboola/db-extractor-adapter": "dev-cftl-771-incremental-fetching-window as 1.15.999",
+        "keboola/db-extractor-common": "dev-cftl-771-incremental-fetching-window as 17.1.999",
+        "keboola/db-extractor-config": "dev-cftl-771-incremental-fetching-window as 1.15.999",
```
[the remark this answers is not in the document]
Comment (matyas-jirat-keboola) [reply composer.json]: This is an intentional, documented interim state. The shared libraries that carry this feature are not released yet (no stable tags exist), so the dev-branch pins mirror the already-merged reference driver. The PR description notes they must be repinned to the released versions and the VCS repositories dropped once the shared-lib release lands — which cannot be done now because those tags do not exist. Resolving.

File: src/Extractor/Hive.php (hunk 1 above)
Comment (matyas-jirat-keboola) [reply src/Extractor/Hive.php]: Fixed in 7db66e7 — a DATE column now returns null from getIncrementalFetchingColumnType(), so window/lookback cleanly reports "not supported by this extractor" for DATE instead of relying on an unverified comparison. Plain watermark on a DATE column is unchanged; DATE window/lookback can be enabled once verified against the live harness. Resolving.

File: tests/functional/incremental-window-late-commit/source/data/config.json (hunk 3 above)
Comment (matyas-jirat-keboola) [reply tests/functional/incremental-window-late-commit/source/data/config.json:19]: Fixed in 7db66e7 — added tests/functional/incremental-window-start-end/, a window case with both incrementalFetchingStart and incrementalFetchingEnd asserting the bounded row set. Resolving.

File: src/Extractor/Hive.php [hunk 5]
```diff
@@ -93,6 +87,38 @@ public function validateIncrementalFetching(ExportConfig $exportConfig): void
...
+    /**
+     * Detects the resolver basetype ("TIMESTAMP" | "INTEGER" | "NUMERIC" | "FLOAT") of the incremental
+     * fetching column. Overrides the default-null hook in db-extractor-common's BaseExtractor, opting
+     * Hive into the incremental fetching window/lookback feature (the type is needed to resolve the
+     * relative/absolute bounds). Reuses the same metadata lookup as validateIncrementalFetching().
+     *
+     * GenericStorage maps a Hive DATE column to the "DATE" basetype, but the window/lookback resolver
+     * only understands TIMESTAMP/INTEGER/NUMERIC/FLOAT, so DATE is normalised to TIMESTAMP here (the
+     * resolver emits "Y-m-d H:i:s", which Hive accepts when comparing a DATE column). Every other type
+     * in self::INCREMENTAL_TYPES already maps to a supported basetype.
```
[the remark this answers is not in the document]
Comment (matyas-jirat-keboola) [reply src/Extractor/Hive.php]: Addressed in 7db66e7 — the docblock no longer asserts the unverified DATE-vs-timestamp behaviour as fact; DATE now returns null (window/lookback is not offered for DATE) rather than documenting an unverified guarantee. Resolving.
